cache/clock_cache.h

```diff
@@ -52,107 +54,173 @@ constexpr double kStrictLoadFactor = 0.7;  // See fast_lru_cache.h.
+  //      * set CLOCK_PRIORITY to NONE;
+  //      * set the HAS_HIT bit.
+  //    Notice that these two types of updates are idempotent, so
+  //    they don't require synchronization across shared references.
+  // - Use an exclusive reference when we want identity members
+  //    to remain untouched, as well as modify any identity member
+  //    or flag.
+  // - displacements can be modified without holding a reference.
+  // - refs is only modified through appropiate functions to
+  //    take or release references.
+
+  ClockHandle()
+      : value(nullptr),
+        deleter(nullptr),
+        hash(0),
+        total_charge(0),
+        refs(0),
+        flags(0),
+        displacements(0) {
+    SetWillDelete(false);
     SetIsElement(false);
     SetClockPriority(ClockPriority::NONE);
     SetCachePriority(Cache::Priority::LOW);
-    displacements = 0;
     key_data.fill(0);
   }
 
-  Slice key() const { return Slice(key_data.data(), kCacheKeySize); }
-
-  // Increase the reference count by 1.
-  void Ref() { refs++; }
-
-  // Just reduce the reference count by 1. Return true if it was last reference.
-  bool Unref() {
-    assert(refs > 0);
-    refs--;
-    return refs == 0;
+  ClockHandle(const ClockHandle& other)
+      : value(other.value),
+        deleter(other.deleter),
+        hash(other.hash),
+        total_charge(other.total_charge),
+        key_data(other.key_data) {
+    refs.store(other.refs);
+    flags.store(other.flags);
+    displacements.store(other.displacements);
+    SetWillDelete(other.WillDelete());
+    SetIsElement(other.IsElement());
+    SetClockPriority(other.GetClockPriority());
+    SetCachePriority(other.GetCachePriority());
   }
 
-  // Return true if there are external refs, false otherwise.
-  bool HasRefs() const { return refs > 0; }
+  void operator=(const ClockHandle& other) {
+    // TODO(Guido) Is there a better way to implement copy ctor + operator=?
+    value = other.value;
+    deleter = other.deleter;
+    hash = other.hash;
+    total_charge = other.total_charge;
+    refs.store(other.refs);
+    key_data = other.key_data;
+    flags.store(other.flags);
+    SetWillDelete(other.WillDelete());
+    SetIsElement(other.IsElement());
+    SetClockPriority(other.GetClockPriority());
+    SetCachePriority(other.GetCachePriority());
+    displacements.store(other.displacements);
+  }
 
-  bool IsVisible() const { return flags & IS_VISIBLE; }
+  Slice key() const { return Slice(key_data.data(), kCacheKeySize); }
 
-  void SetIsVisible(bool is_visible) {
-    if (is_visible) {
-      flags |= IS_VISIBLE;
-    } else {
-      flags &= ~IS_VISIBLE;
-    }
-  }
+  bool HasExternalRefs() const { return (refs & EXTERNAL_REFS) > 0; }
 
   bool IsElement() const { return flags & IS_ELEMENT; }
 
   void SetIsElement(bool is_element) {
     if (is_element) {
       flags |= IS_ELEMENT;
     } else {
-      flags &= ~IS_ELEMENT;
+      flags &= static_cast<uint8_t>(~IS_ELEMENT);
```

Comment:
What's the failure if you don't add "static_cast<uint8_t>"?

Reply:
~ returns a larger data type than uint8_t. Actually only Windows compilation complaints about this.

Reply:
https://github.com/facebook/rocksdb/blob/main/cache/lru_cache.h#L148

This flags is uint8_t. I am not sure why the above line has no issue. Please try and test it.

Reply:
This operator&= from std::atomic<uint8_t>, not uint8_t.

Reply:
Can you try the following?
flags.store(flags & ~IS_ELEMENT);

Reply:
Isn't this going to be compiled as 2 atomic operations on flags?

Reply:
pls ignore my previous comment.

how about flags.fetch_and(~IS_ELEMENT) ?

Reply:
That doesn't work either :(